core/src/main/scala/org/apache/spark/sql/delta/files/DeltaSourceSnapshot.scala

```diff
@@ -52,51 +54,45 @@ class DeltaSourceSnapshot(
     (part, data)
   }
 
-  protected def initialFiles: Dataset[IndexedFile] = {
+  lazy val initialFiles: DataFrame = {
     import spark.implicits.rddToDatasetHolder
     import org.apache.spark.sql.delta.implicits._
 
+    val numPartitions = snapshot.getNumPartitions
+
     cacheDS(
-      snapshot.allFiles.sort("modificationTime", "path")
+      snapshot.allFiles
+        // This allows us to control the number of partitions created from the sort instead of
+        // using the shufflePartitions setting
+        .repartitionByRange(numPartitions, col("modificationTime"), col("path"))
+        .sort("modificationTime", "path")
         .rdd.zipWithIndex()
         .toDF("add", "index")
+        // Stats aren't used for streaming reads right now, so decrease
+        // the size of the files by nulling out the stats if they exist
+        .withColumn("add", col("add").withField("stats", DataSkippingReader.nullStringLiteral))
         .withColumn("remove", SingleAction.nullLitForRemoveFile)
         .withColumn("cdc", SingleAction.nullLitForAddCDCFile)
         .withColumn("version", lit(version))
         .withColumn("isLast", lit(false))
-        .withColumn("shouldSkip", lit(false))
-        .as[IndexedFile],
-      s"Delta Source Snapshot #$version - ${snapshot.redactedPath}").getDS
+        .withColumn("shouldSkip", lit(false)),
+      s"Delta Source Snapshot #$version - ${snapshot.redactedPath}").getDF
   }
 
-  override def close(unpersistSnapshot: Boolean): Unit = {
-    super.close(unpersistSnapshot)
+  def iterator(): Iterator[IndexedFile] = {
+    import org.apache.spark.sql.delta.implicits._
 
-    if (unpersistSnapshot) {
-      snapshot.uncache()
-    }
+    DeltaLog.filterFileList(
+      snapshot.metadata.partitionSchema,
+      initialFiles,
+      partitionFilters,
+      Seq("add")).as[IndexedFile].toLocalIterator.asScala
   }
-}
 
-trait SnapshotIterator {
-  self: DeltaSourceSnapshot =>
-
-  private var result: Iterable[IndexedFile] = _
-
-  def iterator(): Iterator[IndexedFile] = {
-    import org.apache.spark.sql.delta.implicits._
-    if (result == null) {
-      result = DeltaLog.filterFileList(
-        snapshot.metadata.partitionSchema,
-        initialFiles.toDF(),
-        partitionFilters,
-        Seq("add")).as[IndexedFile].collect().toIterable
+  def close(unpersistSnapshot: Boolean): Unit = {
+    uncache()
+    if (unpersistSnapshot) {
+      snapshot.uncache()
     }
-    // This will always start from the beginning and re-use resources. If any exceptions were to
-    // be thrown, the stream would stop, we would call stop on the source, and that will make
-    // sure that we clean up resources.
```

Comment:
Yeah the fact that it's cached should help that. Don't think there's anyway around a two pass method because that's just how building the offsets work. I definitely don't fully understand what that comment is trying to say but its from the initial import of the project there's no real background